tex/testing.tex

```diff
@@ -0,0 +1,33 @@
+
+\section{Feature Integration \& Testing}
+Development of new features naturally means that they must be deemed legal by the scanner, parser, semantic analyzer, and code generator. As we developed new features, the process was roughly as follows:
+  \begin{enumerate}
+    \item Write a simple test that illustrated the feature to test.
+    \item Write the expected output of the aforementioned test to a text file.
+    \item Confirm that the scanner consumes the tokens related to the feature.
+    \item Confirm that the parser grammar has been adjusted to accomodate the new feature.
+    \item Confirm that the semantic analyzer and transformer can properly identify and check the new feature code.
+    \item Confirm that code generation generates the appropriate LLVM IR for the new features - such as allocating memory, building calls, and more.
+    \item Ensure that the test written can write its output to stdout, to be compared with expected output.
+    \item Compile and test the code to ensure that the code has worked to the team's expectations.
+  \end{enumerate}
+
+  \medskip \noindent
+  Earlier in the development process, we tested the front end of our compiler by JSON-ifying the abstract syntax tree and examinining it. As we settled into full-fledged development, we would test with a full-feature regression test suite.
+
+\section{Regression Test Suite}
+Extend's regression test suite is executable through the \texttt{testscript.sh} script at the top level of the project. There are over 100 integration test files for various features of the Extend language, and a corresponding file with their expected output to \texttt{stdout}. This is to ensure that the successful implementation of one feature does not impact that of others.
+
+\medskip \noindent
+The test script iterates through every test file in the \texttt{inputs\_regression} directory. It compiles and executes the test, and compares it with the corresponding expected output file living in the \texttt{expected} directory. Some of these tests were "passing" tests, meant to confirm successful output, and some were "failing" tests, identifying the ability of the language to detect illegal actions and exit gracefully.
+
+\medskip \noindent
+\textbf{Note:} We have added a full test listing at the end of this document. Please refer to the chapter titled "Test Listing" for more detail.
+
+  \subsection{Integration with Travis CI}
+  The aforementioned test suite is run by Travis CI in the event that the Extend compiler is successfully built; otherwise, the build will fail and exit. In our development workflow, checking the logs during build failures sometimes revealed that tests in the regression test suite did not succeed as expected. This integration kept the far-reaching effects of newly introduced features entirely transparent throughout the process.
```

Comment:
I'd write a little more here, mentioning that it was a strong component in keeping our compiler in a working status.